### PR TITLE
[gzhttp] Add supported decompress request body

### PR DESCRIPTION
**Pull request by mirecl**
Add new feature from discussions https://github.com/klauspost/compress/discussions/1001

### CODE REVIEW

File: gzhttp/compress.go [hunk 1]
```diff
@@ -464,6 +464,12 @@ func NewWrapper(opts ...option) (func(http.Handler) http.HandlerFunc, error) {
 	return func(h http.Handler) http.HandlerFunc {
 		return func(w http.ResponseWriter, r *http.Request) {
 			w.Header().Add(vary, acceptEncoding)
+			if contentGzip(r) {
+				if readerGzipBody, err := gzip.NewReader(r.Body); err == nil {
+					r.Body = io.NopCloser(readerGzipBody)
+				}
```
[the remark this answers is not in the document]
Comment (klauspost) [reply gzhttp/compress.go:470]: [zstdReader](https://github.com/klauspost/compress/blob/0deb85a2cb5821965f1d7bf334f4036eb501f06f/gzhttp/transport.go#L200-L204) is also available for easy implementation. (optional)

File: gzhttp/compress.go [hunk 2]
```diff
@@ -464,6 +464,12 @@ func NewWrapper(opts ...option) (func(http.Handler) http.HandlerFunc, error) {
 	return func(h http.Handler) http.HandlerFunc {
 		return func(w http.ResponseWriter, r *http.Request) {
 			w.Header().Add(vary, acceptEncoding)
+			if contentGzip(r) {
+				if readerGzipBody, err := gzip.NewReader(r.Body); err == nil {
+					r.Body = io.NopCloser(readerGzipBody)
```
Comment (klauspost) [inline gzhttp/compress.go:469]: The header field must be removed to avoid upstream confusion.

**Review by klauspost — CHANGES_REQUESTED** [review]
Please add and implement these options:

```
// AllowCompressedRequests will enable or disable RFC 7694 compressed requests.
// By default this is Disabled.
// See https://datatracker.ietf.org/doc/html/rfc7694
func AllowCompressedRequests(b bool) option {
	return func(c *config) {
		//....
	}
}

// SendAcceptEncoding will send Accept-Encoding on all responses as per RFC 7694.
// By default, none will be sent.
// Calling this with no parameters will send the currently supported encodings back.
// This does not affect which encodings are accepted.
func SendAcceptEncoding(enc ...string) option {
	return func(c *config) {
		//....
	}
}
```

While it would be really nice to have this enabled by default, I think the "zip-bomb" risk is too high.

**Comment by klauspost** [comment]
Just as an example [AWS S3](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax) "overloads" `Content-Encoding` on their uploads. in their terms it means the object should remain untouched and always return `Content-Encoding: whatever` when the object is downloaded.  This change would be breaking for that if enabled by default.

File: gzhttp/compress.go [hunk 3]
```diff
@@ -752,6 +758,12 @@ func RandomJitter(n, buffer int, paranoid bool) option {
 	}
 }
 
+// contentGzip returns true if the given HTTP request indicates that it gzipped.
+func contentGzip(r *http.Request) bool {
+	// See more detail in `acceptsGzip`
+	return r.Method != http.MethodHead && parseEncodingGzip(r.Header.Get(contentEncoding)) > 0
```
[the remark this answers is not in the document]
Comment (mirecl) [reply gzhttp/compress.go]: done

File: gzhttp/compress.go (hunk 2 above)
Comment (mirecl) [reply gzhttp/compress.go]: done

**Comment by mirecl** [comment]
> Just as an example [AWS S3](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax) "overloads" `Content-Encoding` on their uploads. in their terms it means the object should remain untouched and always return `Content-Encoding: whatever` when the object is downloaded. This change would be breaking for that if enabled by default.

@klauspost 
+ ``AllowCompressedRequests`` - should help solve this problem
+  ``zstdReader`` - i suggest limiting ourselves to gzip support for now

File: gzhttp/compress.go [hunk 4]
```diff
@@ -464,6 +464,12 @@ func NewWrapper(opts ...option) (func(http.Handler) http.HandlerFunc, error) {
 	return func(h http.Handler) http.HandlerFunc {
 		return func(w http.ResponseWriter, r *http.Request) {
 			w.Header().Add(vary, acceptEncoding)
+			if c.allowCompressedRequests && contentGzip(r) {
+				r.Header.Del(contentEncoding)
+				r.Body = io.NopCloser(&gzipReader{body: r.Body})
```
Comment (klauspost) [inline gzhttp/compress.go:469]: No need to wrap. `*gzipReader` is an `io.ReadCloser` that will close the body. Also it doesn't leak the decoder.

```suggestion
				r.Body = &gzipReader{body: r.Body}
```

File: gzhttp/compress.go [hunk 5]
```diff
@@ -464,6 +464,12 @@ func NewWrapper(opts ...option) (func(http.Handler) http.HandlerFunc, error) {
...
+				defer r.Body.Close()
```
Comment (klauspost) [inline gzhttp/compress.go:470]:
```suggestion
				defer r.Close()
```
With the change above and this the body is always closed and decoder is recycled.

**Review by klauspost — CHANGES_REQUESTED** [review]

File: gzhttp/compress_test.go [hunk 6]
```diff
@@ -89,6 +89,25 @@ func TestMustNewGzipHandler(t *testing.T) {
 	handler.ServeHTTP(res3, req3)
 
 	assertEqual(t, http.DetectContentType([]byte(testBody)), res3.Header().Get("Content-Type"))
+
+	// send compress request body
+	handlerCompressedRequests := newTestHandlerLevel(testBody, AllowCompressedRequests(true))
```
Comment (klauspost) [inline gzhttp/compress_test.go:94]: Please add a test with default settings and check if it remains compressed.

File: gzhttp/compress.go (hunk 5 above)
Comment (mirecl) [reply gzhttp/compress.go:470]: @klauspost, ``defer r.Close`` - not found this function, maybe we should leave the old version ``defer r.Body.Close()``?

Comment (klauspost) [reply gzhttp/compress.go:470]: Ah. Yeah. Correct.

**Comment by mirecl** [comment]
@klauspost, all tasks done! Thank you very much for your patience and tips)

**Comment by klauspost** [comment]
Cool. Giving it a final review! Thanks for the contribution!

**Review by klauspost — APPROVED** [review]

**Pull request merged**